### PR TITLE
Fix various issues found when migrating JoshaberryBush to new code

### PR DESCRIPTION
**Pull request by Josharias**
- Getting the next growth stage is now more array safe
- Harvesting only happens when activating the bush as a character,  this prevents harvesting when using an item
- Remove some complexity on persisting a flat version of growth stages.  This will allow world generated bushes to be added to the world without problems
- Add the cheat items back in the game so that testing new plants is easier
- Let the addition of the plant components force the blocks to be active and persist the added component

### CODE REVIEW

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 1]
```diff
@@ -155,12 +155,31 @@ public void onBushGrowth(DelayedActionTriggeredEvent event, EntityRef bush, Bush
...
+        EntityRef newBush = null;
+        Map.Entry<String, GrowthStage> stage = null;
+        if (!isInLastStage(bushComponent) || stages < 0) {
+            bushComponent.currentStage += stages;
+            stage = getGrowthStage(bushComponent, bushComponent.currentStage);
+            worldProvider.setBlock(position, blockManager.getBlock(stage.getKey()));
+            newBush = blockEntityRegistry.getBlockEntityAt(position);
+            newBush.addOrSaveComponent(bushComponent);
+        }
+        if( newBush != null && stage.getValue().maxTime > 0 && stage.getValue().minTime > 0) {
```
Comment (jellysnake) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: Unless I'm mistaken, the only time `newBush` is not `null` is if the preceding `if` is executed. This should be nested inside the other `if`, also meaning that `newBush` and `stage` don't need to be initialised/declared outside the if statement.

Additionally, handling for the `maxTime` is done in the `generateRandom` method.

**Comment by Josharias** [comment]
@jellysnake, any issues with this set of changes?

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 2]
```diff
@@ -170,25 +189,22 @@ private EntityRef doBushGrowth(BushDefinitionComponent bushComponent, int stages
...
-        EntityRef target = event.getTarget();
-        EntityRef harvester = entity.equals(target) ? event.getInstigator() : entity;
-        if (!event.isConsumed() && areValidHarvestEntities(target, harvester)) {
-
-            BushDefinitionComponent bushComponent = target.getComponent(BushDefinitionComponent.class);
+    public void onHarvest(ActivateEvent event, EntityRef entity, BushDefinitionComponent bushComponent, BlockComponent blockComponent) {
+        EntityRef harvester = event.getInstigator();
+        if (!event.isConsumed() && areValidHarvestEntities(entity, harvester)
+                // only let player characters harvest to avoid item use triggering a harvest
+                && harvester.hasComponent(PlayerCharacterComponent.class)) {
```
Comment (jellysnake) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: This stops functionality like the Harvester Cart in Additional Rails. I think it should be possible for things other than the player to be able to harvest the bush

File: src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java [hunk 3]
```diff
@@ -16,13 +16,11 @@
 package org.terasology.simpleFarming.systems;
 
 import org.terasology.logic.common.ActivateEvent;
-import org.terasology.simpleFarming.components.BushDefinitionComponent;
-import org.terasology.simpleFarming.components.SeedDefinitionComponent;
-import org.terasology.simpleFarming.components.VineNodeComponent;
+import org.terasology.logic.inventory.ItemComponent;
+import org.terasology.simpleFarming.components.*;
```
Comment (jellysnake) [inline src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java]: Star import, shouldn't be used

File: src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java [hunk 4]
```diff
@@ -118,14 +115,32 @@ public void onVinePlanted(OnSeedPlanted event, EntityRef definitionEntity, VineD
...
     @ReceiveEvent
     public void onVineGrowth(DelayedActionTriggeredEvent event, EntityRef root, VineNodeComponent nodeComponent, VineDefinitionComponent vineComponent) {
-        if (nodeComponent.height != -1) {
-            if (nodeComponent.height < 20) {
+        doGrowVine(root, nodeComponent, vineComponent);
+    }
+
+    private void doGrowVine(EntityRef root, VineNodeComponent nodeComponent, VineDefinitionComponent vineComponent) {
+        if (nodeComponent.length != -1) {
+            if (nodeComponent.length < 20) {
```
Comment (jellysnake) [inline src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java]: I noticed that you added a `maxLength` to the `VineDefinitionComponent` but it's not used anywhere. This 20 is the vine's length so replacing it with `maxLength` will implement it.

File: src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java [hunk 5]
```diff
@@ -118,14 +115,32 @@ public void onVinePlanted(OnSeedPlanted event, EntityRef definitionEntity, VineD
...
                 recurseGrow(root, vineComponent);
             }
             resetDelay(root, vineComponent.minGrowTime, vineComponent.maxGrowTime);
         }
     }
 
+    @ReceiveEvent
+    public void onCheatGrowth(ActivateEvent event, EntityRef item, CheatGrowthComponent cheatGrowthComponent, ItemComponent itemComponent) {
```
Comment (jellysnake) [inline src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java]: Can you add some javadoc :)

File: src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java [hunk 6]
```diff
@@ -118,14 +115,32 @@ public void onVinePlanted(OnSeedPlanted event, EntityRef definitionEntity, VineD
...
+        if (!event.getTarget().hasComponent(VineDefinitionComponent.class) || !event.getTarget().hasComponent(VineNodeComponent.class)) {
```
Comment (jellysnake) [inline src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java]: Could you please replace `event.getTarget` with a variable to make it clearer

**Review by syntaxi — CHANGES_REQUESTED** [review]
Immense thanks for adding in the `@ForceBlockActive` which is exactly what I managed to miss and needed.

I'm dubious about restricting the harvest to only player as that cuts off any form of automation in the future, most notably the current Harvest Cart at the moment.
Right click doesn't harvest anyway, `E` must be used.

I've no problems with solely using the Map for growth stages, although it still doesn't allow for bushes to be created in world gen (at least ones that will grow correctly to the best of my knowledge) as it still requires  entity creation off thread.

Other than that the always present Javadoc (But that's not overly a blocker)

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java (hunk 1 above)
Comment (Josharias) [reply src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: Yes!  Good insight.

I have not seen handling of maxTime when both max and min are 0 (which happens if they are not specified in json).  Maybe there is a better way to address this than the double >0 check?

File: src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java (hunk 3 above)
Comment (Josharias) [reply src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java]: Thanks,  I have fixed my IDE settings and un-starred these.

File: src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java (hunk 4 above)
Comment (Josharias) [reply src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java]: Hah,  I didnt quite finish implementing that one.  Doh.

Also,  @flo and I made some long vines yesterday with the growthcan...  much fun was had!

File: src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java (hunk 6 above)
Comment (Josharias) [reply src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java]: Deal.

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java (hunk 2 above)
Comment (Josharias) [reply src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java]: You are right.  I had some other problem that caused this behavior.  Thanks for the sanity check.  <removed>

**Comment by jellysnake** [comment]
This is ready to be merged then :)

**Comment by Josharias** [comment]
Thank you for the code review, @jellysnake

**Pull request merged**